FR #956: Provider-calculated split cost allocation (#1394)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/allocatedtags.md

# Allocated Tags

The Allocated Tags column represents the set of [*tags*](#glossary:tag) assigned to [*tag sources*](#glossary:tag-source) which are specifically applicable to allocated charges resulting from a provider-calculated split cost allocation.

The AllocatedTags column adheres to the following requirements:

* AllocatedTags MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports provider-calculated split cost allocation.
* AllocatedTags MUST conform to [KeyValueFormat](#key-valueformat) requirements.

> * AllocatedTags MAY be null.

Reviewer: As mentioned in [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1394/files#r2318478128), I assume the following applies to all Allocated-* columns except AllocatedMethodId:

- [AllocatedColumn] MUST be null when a *charge* represents the unallocated portion of the origin *charge* after split cost allocation.

```suggestion
* AllocatedTags nullability is defined as follows:
  * AllocatedTags MUST be null when a *charge* is not related to a provider-calculated split cost allocation.
  * AllocatedTags MUST be null when a *charge* represents the unallocated portion of the origin *charge* after split cost allocation.
  * AllocatedTags MAY be null in all other cases.
```

Author: Good callout, I used Tags as a foundation for this and didn't update the verbiage of this line.

I'm not sure we should force this to be null if it's unallocated. Google uses their equivalent of AllocatedTags to communicate information about the unallocated portion to differentiate the multiple things that fall into this category.

Author: Made the change, less the unallocated tags piece. Leaving thread open for discussion.

Reviewer 2: @ijurica Can you please review this thread and let @kk09v if we can close it?

Reviewer: Discussed in TF-2 and in an ad-hoc meeting on Wed, Sep 3.
Agreed that current version is just fine.

> * AllocatedTags nullability is defined as follows:
>   * AllocatedTags MUST be null when a *charge* is not related to a provider-calculated split cost allocation.
>   * AllocatedTags MAY be null in all other cases.

Resolving this conversation.

> # Allocated Tags

Reviewer 3: I'm not going to block on this, but I do not think we should add this column. This column requires practitioners to manage 2 sets of tags. Tags are difficult enough with 1 column. This will make a painful process even more painful with unnecessary complexity. I would rather use Tags for the allocated tags.

Reviewer 4: We could maybe put allocated tags in the existing tag column for allocation records only, but I think that adds another level of complexity for the practitioner having to potentially join the tags found in the allocated record to the original charge tags.

@kk09v what say you?

Reviewer 4: Discussed in TF2, majority agreed to leave as is.

> The Allocated Tags column represents the set of [*tags*](#glossary:tag) assigned to [*tag sources*](#glossary:tag-source) which are specifically applicable to allocated charges resulting from a provider-calculated split cost allocation.

Reviewer 3: Same comment as earlier: I would really like to avoid the jargon here if we can.
```suggestion
The Allocated Tags column represents the set of [*tags*](#glossary:tag) assigned to [*tag sources*](#glossary:tag-source) which are specifically applicable to charges that have been allocated by the provider.
```

Reviewer 4: Noted and changed to "that have been calculated and allocated by the provider." which isn't exactly your edit, but I think works?

The AllocatedTags column adheres to the following requirements:

> * AllocatedTags MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports provider-calculated split cost allocation.

Reviewer 3: Same jargon simplification...
```suggestion
* AllocatedTags MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports splitting resource usage into multiple records based on allocation properties.
```

Reviewer 4: I think in this case, even though it sounds jargon-y, it fits because provider-calculated split cost allocation is the name of the supported feature. Do you still think it's too much and we don't need to be so precise?

  * Provider SHOULD publish all provider-specified tag key prefixes within their respective documentation.
* User-defined tags adhere to the following additional requirements:
  * Provider MUST prefix all user-defined tags scheme with a predetermined, provider-specified tag key prefix that is unique to each corresponding user-defined tag scheme when the provider has more than one user-defined tag scheme.

## Provider-Defined vs. User-Defined Tags

This example illustrates various tags produced from multiple user-defined and provider-defined tag schemes. The first two tags illustrate examples from two different, user-defined tag schemes. The second tag is produced from a valueless, user-defined tag scheme, so the provider also applies `true` as its default value.

The last two tags illustrate examples from two different, provider-defined tag schemes.

```json
    {
        "userDefinedTagScheme1/foo": "bar",
        "userDefinedTagScheme2/foo": true,
        "providerDefinedTagScheme1/foo": "bar",
        "providerDefinedTagScheme2/foo": "bar"
    }
```

## Column ID

AllocatedTags

## Display Name

Allocated Tags

## Description

> The Allocated Tags column represents the set of tags assigned to tag sources which are specifically applicable to *allocated charges* that have been calculated and allocated by the provider.

Reviewer 2: ```suggestion
A set of tags assigned to tag sources that are applicable to *allocated charges* in provider-defined split cost allocation.
```
